### ISSUE TITLE
napari sometimes not starting event loop from CLI if IPython console was left open.

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#2925]
## 🐛 Bug
just came across a strange and hard to figure out bug... still don't fully understand.

When I tried to open a file from the CLI with `napari --plugin some_plugin path/to/file`, napari would get all the way to the `run()` command, but then do nothing.  turned out that run was getting called, but it thought that ipython already had the event loop, so did nothing and quit.  This only happened because I had left the console open in the previous run, so the settings tried to reopen it (which brings IPython into sys.modules ... and it's possible the bioformats/pims plugin is also involved here)?

Long story short.  We should probably make sure to "force" the event loop to start if it's getting called from `napari.__main__` ... and I might even suggest that we always hide the embedded console on startup/shutdown.

**Comment by tlambert03** [comment]
want to add to this:
I think what @JoOkuma observed in https://github.com/napari/napari/issues/3033 is related to this.
I think that our `_try_enable_ipython_gui` is too greedy.  It will look for IPython in `sys.modules`, and if it finds it, attempt to run `shell.enable_gui`.
Because _our own_ console imports IPython, this means that the second call to `get_app` in a single session (For instance, the second creation of a viewer) will actually try to enable `%gui qt`... meaning that additional calls to `napari.run()` will likely not have the blocking effect that is intended in a non IPython environment.

moral: we need a better way to determine that we're actually in an ipython interactive session (rather than just using ipython within napari while running in a script, or `napari.__main__`)

**Comment by tlambert03** [comment]
@Carreau, I'm not quite sure what the best way to detect this would be.  It would be easy enough to check `sys.argv`, but I don't know whether that would work in, for example, a jupyter environment.

Can you think of a good way to differentiate between napari-console creating an InteractiveShell within an already existing python process started directly from a script (or from `python -m napari`), vs someone _starting_ in IPython/Jupyter and then importing napari and creating a viewer?
